src/lib/y2storage/proposal/autoinst_devices_planner.rb

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2017] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2storage/disk"
+
+module Y2Storage
+  module Proposal
+    # Class to generate a list of Planned::Device objects that must be allocated
+    # during the AutoYaST proposal.
+    class AutoinstDevicesPlanner
+      include Yast::Logger
+
+      # Constructor
+      #
+      # @param devicegraph [Devicegraph] Devicegraph to be used as starting point
+      def initialize(devicegraph)
+        @devicegraph = devicegraph
+      end
+
+      # Returns an array of planned devices according to the drives map
+      #
+      # @raise [Error] if not partitions were specified
+      #
+      # @param drives_map [Proposal::AutoinstDrivesMap] Drives map from AutoYaST
+      # @return [Array<Planned::Partition>] List of planned partitions
+      def planned_devices(drives_map)
+        result = []
+
+        drives_map.each_pair do |disk_name, drive_spec|
```

Comment:
A matter of taste, most people from the storage team prefer the existing alternative to the `reduce`-base one.